File: lint-arwaky/src-python/taxonomy/metrics_provider_error.py
"""metrics_provider_error — Technical metrics domain error types."""

from pydantic import BaseModel, ConfigDict

from .error_code_vo import ErrorCode
from .error_value_vo import (
    ErrorMessage,
    Cause,
)
from .file_path_vo import FilePath


class MetricsError(BaseModel):
    """General failure during metrics retrieval."""

    model_config = ConfigDict(frozen=True)

    path: FilePath | None = None
    message: ErrorMessage
    error_code: ErrorCode | None = None
    cause: Cause | None = None

    def __str__(self):
        target = f" for {self.path}" if self.path else ""
        code = f" [{self.error_code}]" if self.error_code else ""
        return f"Metrics Error{target}{code}: {self.message}"

File: lint-arwaky/src-rust/taxonomy/metrics_provider_error.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct MetricsError {
    #[serde(default)]
    pub path: Option<FilePath>,
    pub message: ErrorMessage,
    #[serde(default)]
    pub error_code: Option<ErrorCode>,
    #[serde(default)]
    pub cause: Option<Cause>,
}

impl MetricsError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { path: None, message, error_code: None, cause: None }
    }
}

impl std::fmt::Display for MetricsError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let target = self.path.as_ref().map(|p| format!(" for {}", p)).unwrap_or_default();
        let code = self.error_code.as_ref().map(|c| format!(" [{}]", c)).unwrap_or_default();
        write!(f, "Metrics Error{}{}: {}", target, code, self.message)
    }
}
